File: aoc-2025/rust/src/bin/day08.rs
//! https://adventofcode.com/2025/day/8
use std::env;
use std::fs;
use std::path::Path;

const DEFAULT_INPUT: &str = "../input/day08.txt";
const DEFAULT_OUTPUT: &str = "output/day08.txt";

#[derive(Clone, Copy, Debug)]
struct Point {
    x: i64,
    y: i64,
    z: i64,
}

#[derive(Clone, Copy, Debug)]
struct Edge {
    u: usize,
    v: usize,
    dist_sq: i64,
}

struct Dsu {
    parent: Vec<usize>,
    size: Vec<usize>,
    comp_cnt: usize,
}

impl Dsu {
    fn new(n: usize) -> Self {
        Dsu {
            parent: (0..n).collect(),
            size: vec![1; n],
            comp_cnt: n,
        }
    }

    fn find(&mut self, i: usize) -> usize {
        let mut root = i;
        while root != self.parent[root] {
            root = self.parent[root];
        }
        let mut curr = i;
        while curr != root {
            let next = self.parent[curr];
            self.parent[curr] = root;
            curr = next;
        }
        root
    }

    fn unite(&mut self, i: usize, j: usize) -> bool {
        let root_i = self.find(i);
        let root_j = self.find(j);
        if root_i != root_j {
            if self.size[root_i] < self.size[root_j] {
                self.parent[root_i] = root_j;
                self.size[root_j] += self.size[root_i];
            } else {
                self.parent[root_j] = root_i;
                self.size[root_i] += self.size[root_j];
            }
            self.comp_cnt -= 1;
            true
        } else {
            false
        }
    }
}

fn parse_points(content: &str) -> Vec<Point> {
    content
        .lines()
        .filter(|line| !line.trim().is_empty())
        .map(|line| {
            let parts: Vec<i64> = line
                .replace(',', " ")
                .split_whitespace()
                .filter_map(|s| s.parse().ok())
                .collect();

            Point {
                x: parts[0],
                y: parts[1],
                z: parts[2],
            }
        })
        .collect()
}

fn generate_edges(points: &[Point]) -> Vec<Edge> {
    let n = points.len();
    if n == 0 {
        return Vec::new();
    }
    let mut edges = Vec::with_capacity(n * (n - 1) / 2);
    for i in 0..n {
        for j in (i + 1)..n {
            let dx = points[i].x - points[j].x;
            let dy = points[i].y - points[j].y;
            let dz = points[i].z - points[j].z;
            let dist_sq = dx * dx + dy * dy + dz * dz;
            edges.push(Edge {
                u: i,
                v: j,
                dist_sq,
            });
        }
    }
    edges
}

fn part_one(content: &str) -> i64 {
    let points = parse_points(content);
    if points.is_empty() {
        return 0;
    }

    let mut edges = generate_edges(&points);
    edges.sort_unstable_by_key(|e| e.dist_sq);

    let mut dsu = Dsu::new(points.len());

    let k = 1000.min(edges.len());
    for edge in edges.iter().take(k) {
        dsu.unite(edge.u, edge.v);
    }

    let mut sizes: Vec<i64> = (0..points.len())
        .filter(|&i| dsu.parent[i] == i)
        .map(|i| dsu.size[i] as i64)
        .collect();

    sizes.sort_unstable_by(|a, b| b.cmp(a));

    sizes.iter().take(3).product()
}

fn part_two(content: &str) -> i64 {
    let points = parse_points(content);
    if points.is_empty() {
        return 0;
    }

    let mut edges = generate_edges(&points);
    edges.sort_unstable_by_key(|e| e.dist_sq);

    let mut dsu = Dsu::new(points.len());

    for edge in edges {
        if dsu.unite(edge.u, edge.v) && dsu.comp_cnt == 1 {
            return points[edge.u].x * points[edge.v].x;
        }
    }

    0
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_path = args.get(1).map(|s| s.as_str()).unwrap_or(DEFAULT_INPUT);
    println!("Reading from: {}", input_path);

    let content = fs::read_to_string(input_path).expect("Could not read input file");

    let p1 = part_one(&content);
    let p2 = part_two(&content);

    println!("Part 1: {}", p1);
    println!("Part 2: {}", p2);

    if let Some(parent) = Path::new(DEFAULT_OUTPUT).parent() {
        let _ = fs::create_dir_all(parent);
    }
    let _ = fs::write(DEFAULT_OUTPUT, format!("{}\n{}\n", p1, p2));
}

File: aoc-2025/cpp/src/day08.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <vector>

const std::string DEFAULT_INPUT = "";
const std::string DEFAULT_OUTPUT = "";

struct Point {
  long long x, y, z;
};

struct Edge {
  int u, v;
  long long dist_sq;
};

struct DSU {
  std::vector<int> parent;
  std::vector<long long> sz;
  int comp_cnt;

  DSU(int n) {
    parent.resize(n);
    std::iota(parent.begin(), parent.end(), 0);
    sz.assign(n, 1);
    comp_cnt = n;
  }

  int find(int i) {
    if (parent[i] == i)
      return i;
    return parent[i] = find(parent[i]);
  }

  bool unite(int i, int j) {
    int root_i = find(i);
    int root_j = find(j);
    if (root_i != root_j) {
      if (sz[root_i] < sz[root_j])
        std::swap(root_i, root_j);
      parent[root_j] = root_i;
      sz[root_i] += sz[root_j];
      comp_cnt--;
      return true;
    }
    return false;
  }
};

std::vector<Point> read_points(const std::string &filename) {
  std::ifstream file(filename);
  std::vector<Point> points;
  if (!file.is_open())
    return points;

  std::string line;
  while (std::getline(file, line)) {
    if (line.empty())
      continue;
    for (char &c : line)
      if (c == ',')
        c = ' ';
    std::stringstream ss(line);
    long long x, y, z;
    if (ss >> x >> y >> z)
      points.push_back({x, y, z});
  }
  return points;
}

std::vector<Edge> generate_edges(const std::vector<Point> &points) {
  std::vector<Edge> edges;
  int n = points.size();
  edges.reserve((size_t)n * (n - 1) / 2);

  for (int i = 0; i < n; ++i)
    for (int j = i + 1; j < n; ++j) {
      long long dx = points[i].x - points[j].x;
      long long dy = points[i].y - points[j].y;
      long long dz = points[i].z - points[j].z;
      long long dist_sq = dx * dx + dy * dy + dz * dz;
      edges.push_back({i, j, dist_sq});
    }
  return edges;
}

long long part_one(const std::string &filename) {
  auto points = read_points(filename);
  if (points.empty())
    return -1;

  auto edges = generate_edges(points);

  std::sort(edges.begin(), edges.end(),
            [](const Edge &a, const Edge &b) { return a.dist_sq < b.dist_sq; });

  DSU dsu(points.size());

  int k = std::min(1000, (int)edges.size());
  for (size_t i = 0; i < k; ++i)
    dsu.unite(edges[i].u, edges[i].v);

  std::vector<long long> component_sizes;
  for (int i = 0; i < points.size(); ++i)
    if (dsu.parent[i] == i)
      component_sizes.push_back(dsu.sz[i]);

  std::sort(component_sizes.rbegin(), component_sizes.rend());

  long long ans = 1;
  for (int i = 0; i < 3 && i < (int)component_sizes.size(); ++i)
    ans *= component_sizes[i];
  return ans;
}

long long part_two(const std::string &filename) {
  auto points = read_points(filename);
  if (points.empty())
    return -1;

  auto edges = generate_edges(points);

  std::sort(edges.begin(), edges.end(),
            [](const Edge &a, const Edge &b) { return a.dist_sq < b.dist_sq; });

  DSU dsu(points.size());

  for (const auto &edge : edges)
    if (dsu.unite(edge.u, edge.v))
      if (dsu.comp_cnt == 1)
        return points[edge.u].x * points[edge.v].x;

  return 0;
}

int main(int argc, char *argv[]) {
  std::string input_file = (argc > 1) ? argv[1] : DEFAULT_INPUT;
  std::cout << "Reading from: " << input_file << std::endl;

  long long p1 = part_one(input_file);
  long long p2 = part_two(input_file);

  std::cout << "Part 1: " << p1 << std::endl;
  std::cout << "Part 2: " << p2 << std::endl;

  std::ofstream oFile(DEFAULT_OUTPUT);
  if (oFile.is_open())
    oFile << p1 << "\n" << p2 << "\n";

  return 0;
}
